File: src/utils/deepMap.ts
class DeepMapEntry<K, V> {
    value: V | undefined
    isInitialized: boolean = false
    children: Map<K, DeepMapEntry<K, V>> = new Map()

    entries(prefix: K[]): [K[], V][] {
        const children = [...this.children.entries()]
            .flatMap(([key, value]) => {
                return value.entries([...prefix, key])
            })

        if (this.isInitialized) {
            const current: [K[], V] = [prefix, this.value as V]
            return [current, ...children]
        } else {
            return children
        }
    }

    keys(prefix: K[]): K[][] {
        return this.entries(prefix).map(([key]) => key)
    }

    values(prefix: K[]): V[] {
        return this.entries(prefix).map(([, value]) => value)
    }
}

export class DeepMap<K, V> {
    private root: DeepMapEntry<K, V> = new DeepMapEntry()

    clear(): void {
        this.root = new DeepMapEntry()
    }

    delete(key: K[]): void {
        let current: DeepMapEntry<K, V> = this.root

        for (const keyItem of key) {
            const child = current.children.get(keyItem)
            if (child === undefined) return
            current = child
        }

        current.value = undefined
        current.isInitialized = false
    }

    get(key: K[]): V | undefined {
        let current: DeepMapEntry<K, V> = this.root

        for (const keyItem of key) {
            const child = current.children.get(keyItem)
            if (child === undefined) return
            current = child
        }

        return current.value
    }

    has(key: K[]): boolean {
        let current: DeepMapEntry<K, V> = this.root

        for (const keyItem of key) {
            const child = current.children.get(keyItem)
            if (child === undefined) return false
            current = child
        }

        return current.isInitialized
    }

    set(key: K[], value: V): void {
        let current: DeepMapEntry<K, V> = this.root

        for (const keyItem of key) {
            let child = current.children.get(keyItem)
            if (child === undefined) {
                child = new DeepMapEntry()
                current.children.set(keyItem, child)
            }
            current = child
        }

        current.value = value
        current.isInitialized = true
    }

    entries(): [K[], V][] {
        return this.root.entries([])
    }

    keys(): K[][] {
        return this.root.keys([])
    }

    values(): V[] {
        return this.root.values([])
    }
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/util/DeepMap.kt
package io.github.sgrishchenko.karakum.util

import js.array.JsTuple2
import js.array.ReadonlyArray
import js.array.tupleOf

private class DeepMapEntry<K, V> {
    var value: V? = null
    var isInitialized = false
    val children = mutableMapOf<K, DeepMapEntry<K, V>>()

    fun entries(prefix: ReadonlyArray<K>): ReadonlyArray<JsTuple2<ReadonlyArray<K>, V>> {
        val children = children.flatMap { (key, value) ->
            value.entries(prefix + key).asIterable()
        }.toTypedArray()

        if (isInitialized) {
            @Suppress("UNCHECKED_CAST")
            val current = tupleOf(prefix, value as V)
            return arrayOf(current) + children
        } else {
            return children
        }
    }

    fun keys(prefix: ReadonlyArray<K>): ReadonlyArray<ReadonlyArray<K>> {
        return entries(prefix).map { (key) -> key }.toTypedArray()
    }

    fun values(prefix: ReadonlyArray<K>): ReadonlyArray<V> {
        return entries(prefix).map { (_, value) -> value }.toTypedArray()
    }
}

@OptIn(ExperimentalJsExport::class)
@JsExport
class DeepMap<K, V> {
    private var root = DeepMapEntry<K, V>()

    fun clear() {
        root = DeepMapEntry()
    }

    fun delete(key: ReadonlyArray<K>) {
        var current = root

        for (keyItem in key) {
            val child = current.children[keyItem] ?: return
            current = child
        }

        current.value = null
        current.isInitialized = false
    }

    @JsExport.Ignore
    operator fun minusAssign(key: ReadonlyArray<K>) = delete(key)

    operator fun get(key: ReadonlyArray<K>): V? {
        var current = root

        for (keyItem in key) {
            val child = current.children[keyItem] ?: return null
            current = child
        }

        return current.value
    }

    fun has(key: ReadonlyArray<K>): Boolean {
        var current = root

        for (keyItem in key) {
            val child = current.children[keyItem] ?: return false
            current = child
        }

        return current.isInitialized
    }

    @JsExport.Ignore
    operator fun contains(key: ReadonlyArray<K>): Boolean = has(key)

    operator fun set(key: ReadonlyArray<K>, value: V) {
        var current = root

        for (keyItem in key) {
            var child = current.children[keyItem]
            if (child == null) {
                child = DeepMapEntry()
                current.children[keyItem] = child
            }
            current = child
        }

        current.value = value
        current.isInitialized = true
    }

    @JsExport.Ignore
    operator fun plusAssign(pair: Pair<ReadonlyArray<K>, V>) = set(pair.first, pair.second)

    fun entries() = root.entries(emptyArray())

    fun keys() = root.keys(emptyArray())

    fun values() = root.values(emptyArray())
}
